datadog/resource_datadog_synthetics_test_.go

```diff
@@ -3783,3 +3791,71 @@ func validateSyntheticsAssertionOperator(val interface{}, key string) (warns []s
 	}
 	return
 }
+
+func getConfigCertificate(rawConfig cty.Value, stepIndex int) (*string, *string) {
+
+	basePath := cty.GetAttrPath("api_step").
+		Index(cty.NumberIntVal(int64(stepIndex))).
+		GetAttr("request_client_certificate").
+		Index(cty.NumberIntVal(0))
+
+	// Construct paths to the cert and key content
+	certPath := basePath.
+		GetAttr("cert").
+		Index(cty.NumberIntVal(0)).
+		GetAttr("content")
+
+	keyPath := basePath.
+		GetAttr("key").
+		Index(cty.NumberIntVal(0)).
+		GetAttr("content")
+
+	// Apply paths to retrieve the cert and key
+	certValue, err := certPath.Apply(rawConfig)
+	if err != nil || !certValue.IsKnown() || certValue.IsNull() {
+		return nil, nil
+	}
+
+	keyValue, err := keyPath.Apply(rawConfig)
+	if err != nil || !keyValue.IsKnown() || keyValue.IsNull() {
+		return nil, nil
+	}
+
+	// Convert certValue and keyValue to strings
+	certString := certValue.AsString()
+	keyString := keyValue.AsString()
+
+	return &certString, &keyString
+}
+
+func overrideStateCertificate(requestClientCertificates []interface{}, configCert, configKey string) error {
+
+	if len(requestClientCertificates) == 0 {
+		return fmt.Errorf("requestClientCertificates is empty")
+	}
+	requestClientCertificate, ok := requestClientCertificates[0].(map[string]interface{})
+	if !ok {
+		return fmt.Errorf("requestClientCertificates[0] is not a map")
+	}
+	certList, ok := requestClientCertificate["cert"].([]interface{})
+	if !ok || len(certList) == 0 {
+		return fmt.Errorf("cert is not a valid list or is empty")
+	}
+	cert, ok := certList[0].(map[string]interface{})
+	if !ok {
+		return fmt.Errorf("cert[0] is not a map")
+	}
+	cert["content"] = configCert
+
+	keyList, ok := requestClientCertificate["key"].([]interface{})
+	if !ok || len(keyList) == 0 {
+		return fmt.Errorf("key is not a valid list or is empty")
+	}
+	key, ok := keyList[0].(map[string]interface{})
+	if !ok {
+		return fmt.Errorf("key[0] is not a map")
+	}
+	key["content"] = configKey
+
+	return nil
+}
```

Comment:
Mutation is to be avoided as much as possible. Instead, let's have this function return the `requestClientCertificates` interface, and override it in the caller.
```suggestion
func buildDatadogRequestClientCertificate(configCert, configKey string)  {
	requestClientCertificates := datadogV1. ... TODO

	certList, ok := requestClientCertificate["cert"].([]interface{})
	cert, ok := certList[0].(map[string]interface{})
	if !ok {
		return fmt.Errorf("cert[0] is not a map")
	}
	cert["content"] = configCert

	keyList, ok := requestClientCertificate["key"].([]interface{})
	if !ok || len(keyList) == 0 {
		return fmt.Errorf("key is not a valid list or is empty")
	}
	key, ok := keyList[0].(map[string]interface{})
	if !ok {
		return fmt.Errorf("key[0] is not a map")
	}
	key["content"] = configKey

	return requestClientCertificates
}
```

Reply:
Agreed that we should avoid the mutation. However, `completeSyntheticsTestRequest` is not expecting a nicely formatted `requestClientCertificates`, but rather a raw `stepMap["request_client_certificate"]` with a more complexe structure. So the `buildDatadogRequestClientCertificate` does not really make sense, even more that we already have the `buildDatadogRequestCertificates` function which kinda does the same thing, inside of `completeSyntheticsTestRequest`.